### PR TITLE
Bug(#1542) + security =  reDos security and optimisation. + Bug on Ipfs extractCid fixed and opti

### LINKED ISSUE #1542: (JS-S001) Regular Expression Denial of Service (ReDoS)
## Description
The Regular expression Denial of Service (ReDoS) is a Denial of Service attack that exploits the fact that most Regular Expression implementations may reach extreme situations that cause them to work very slowly (exponentially related to input size).

## Occurrences
There are 4 occurrences of this issue in the repository.

See all occurrences on DeepSource &rarr; [deepsource.io/gh/kodadot/nft-gallery/issue/JS-S001/occurrences/](https://deepsource.io/gh/kodadot/nft-gallery/issue/JS-S001/occurrences/)

image

### PR DESCRIPTION
**Pull request by MSghais**
**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [X] Bugfix
- [ ] Feature
- [ ] Refactoring

### What's new?

- [X] PR closes #1542 
- [X] Check reDos security and opti regex on : 
utils/extension.ts 
rmrk/service/NftUtils.ts
rmrk/utils

 /!\  :  utils/ipfs .ts
 Opti + Bug on Ipfs extractCid fixed : change the RegEx and add others check

### Before submitting Pull Request, please make sure:

- [X] My contribution builds **clean without any errors or warnings**
- [X] I've merged recent default branch -- **main** and I've no conflicts
- [X] I've tried to respect high code quality standards
- [X] I've didn't break any original functionality
- [X] I've posted a screenshot of demonstrated change in this PR

### Optional

- [X] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [X] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### Had issue bounty label?

- [X] Fill up your KSM address: [Payout](https://beta.kodadot.xyz/transfer/?target=EFuhGSJpetTgoZ2dfWxwirxADBWgAzWmVfJmTCauZuPxz6z

### Community participation

- [X] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

### Screenshot

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.

### CODE REVIEW

**Review by petersopko — CHANGES_REQUESTED** [review]
can you please check the deepsource issues and see if it's fixable?

**Comment by MSghais** [comment]
> can you please check the deepsource issues and see if it's fixable?

Hi yes the deepsource issues of performance and security is fixed now.

**Review by MSghais — COMMENTED** [review]
Self review. All deepsources issues of performance and security are fixed.

1. Change str.match() to Regex.search.
2. ipfs.ts => Change Regex Ipfs extractCid, fix perf issues and prevent ReDos.

image

File: components/rmrk/service/NftUtils.ts [hunk 1]
```diff
@@ -278,7 +278,7 @@ class NFTUtils {
 
   public static unwrap(rmrkString: string): any {
```
Comment (vikiival) [inline components/rmrk/service/NftUtils.ts:279]: this was moved to minimark afaik

File: components/rmrk/utils.ts [hunk 2]
```diff
@@ -256,7 +256,7 @@ export const getView = (rmrkString: string): RmrkMint | RmrkView | null => {
 
 export const unwrap = (rmrkString: string): any | null => {
```
Comment (vikiival) [inline components/rmrk/utils.ts:257]: this was moved to minimark afaik

File: utils/ipfs.ts [hunk 3]
```diff
@@ -9,23 +9,28 @@ export const justHash = (ipfsLink?: string): boolean => {
...
 
-  const match = ipfsLink.match(cidRegex)
-
-  return match ? match[1] : ''
+  return ipfsLink.replace('ipfs://ipfs/', '')
 }
 
-export const fastExtract = (ipfsLink?: string): string => {
+const cidRegex = /ipfs\/([a-zA-Z0-9]+)/
+export const extractCid = (ipfsLink?: string): string => {
```
Comment (vikiival) [inline utils/ipfs.ts:21]: maybe suitable to write unit test here :)

**Review by vikiival — CHANGES_REQUESTED** [review]

File: utils/ipfs.ts (hunk 3 above)
Comment (MSghais) [reply utils/ipfs.ts:21]: done in ipfsUtils.spect.ts : could extractCid

**Review by MSghais — COMMENTED** [review]
Unit test done for ipfsUtils.spec.ts

**Review by MSghais — COMMENTED** [review]
Feedback done

**Comment by vikiival** [comment]
You still have unresolved item in the conversation

**Comment by vikiival** [comment]
Looks like you merged #2995 here

**Comment by vikiival** [comment]
Please make a new PR with correct code

**Pull request closed without merging**